CMakeLists.txt

```diff
@@ -1,74 +1,105 @@
+# take care of the CMAKE_OSX_SYSROOT. The variable is ignored on non-Apple systems.
 
-if(NOT CURA_ARCUS_BRANCH_OR_TAG)
-    set(CURA_ARCUS_BRANCH_OR_TAG "master")
-endif()
-if(NOT CURA_SAVITAR_BRANCH_OR_TAG)
-    set(CURA_SAVITAR_BRANCH_OR_TAG "master")
-endif()
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake)
+include(GetFromEnvironmentOrCache)
+GetFromEnvironmentOrCache(
+        NAME
+            CMAKE_OSX_DEPLOYMENT_TARGET
+        DESCRIPTION
+            "Specify the minimum version of the target platform (e.g. macOS or iOS)")
 
-# This should care that our CMAKE_INSTALL_PREFIX is absolute at the end...
-get_filename_component(CMAKE_INSTALL_PREFIX
-                       ${CMAKE_INSTALL_PREFIX}
-                       ABSOLUTE
-                       CACHE FORCE)
+project(cura-build-environment)
+cmake_minimum_required(VERSION 3.20)
 
-list(APPEND CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake)
+include(cmake/conan.cmake)
+include(cmake/os.cmake)
 
 include(ExternalProject)
 include(GNUInstallDirs)
+include(ProcessorCount)
 
-include(BuildPlatformDetection)
-include(SetProjectDependencies)
+ProcessorCount(N_PROC)
+if(N_PROC EQUAL 0)
+    set(N_PROC 1)
+endif()
 
-# Hard-code the CPython executable to use later because all find_package()
-# commands are executed when the Makefiles are generated. Because we need to
-# compile CPython first, all other projects that require Python must use our
-# compiled one. Because we cannot use find_package() to find Python at runtime,
-# we have to hard-code the CPython executable that will be created later. All
-# projects that requires Python must use the variable Python3_EXECUTABLE so they
-# will use the right CPython.
-if(BUILD_OS_WINDOWS)
-    # Minimalistic preinstalled Python on Windows:
-    set(Python3_EXECUTABLE ${CMAKE_INSTALL_PREFIX}/bin/python.exe)
-else()
-    # All other OSs like OSX
-    set(Python3_EXECUTABLE ${CMAKE_INSTALL_PREFIX}/bin/python3)
+# Create the version-related variables
+GetFromEnvironmentOrCache(
+        NAME
+            CURA_VERSION_MAJOR
+        DEFAULT
+            "5"
+        DESCRIPTION
+            "Cura Major Version")
+GetFromEnvironmentOrCache(
+        NAME
+            CURA_VERSION_MINOR
+        DEFAULT
+            "0"
+        DESCRIPTION
+            "Cura Minor Version")
+GetFromEnvironmentOrCache(
+        NAME
+            CURA_VERSION_PATCH
+        DEFAULT
+            "0"
+        DESCRIPTION
+            "Cura Patch Version")
+GetFromEnvironmentOrCache(
+        NAME
+            CURA_VERSION_PRE_RELEASE_TAG
+        DESCRIPTION
+            "Cura pre release tag Version Information a, b or rc")
+GetFromEnvironmentOrCache(
+        NAME
+            CURA_VERSION_BUILD
+        DESCRIPTION
+            "Cura Version build")
+set(CURA_VERSION_EXTRA )
+if(NOT ${CURA_VERSION_PRE_RELEASE_TAG} STREQUAL "")
+    set(CURA_VERSION_EXTRA "-${CURA_VERSION_PRE_RELEASE_TAG}+${CURA_VERSION_BUILD}")
```

Comment:
The build metadata tag in SemVer is most commonly a increasing version number. But after reading https://semver.org/ 
> Build metadata MAY be denoted by appending a plus sign and a series of dot separated identifiers immediately following the patch or pre-release version. Identifiers MUST comprise only ASCII alphanumerics and hyphens [0-9A-Za-z-]. Identifiers MUST NOT be empty. Build metadata MUST be ignored when determining version precedence. Thus two versions that differ only in the build metadata, have the same precedence. Examples: 1.0.0-alpha+001, 1.0.0+20130313144700, 1.0.0-beta+exp.sha.5114f85, 1.0.0+21AF26D3—-117B344092BD.

And since we now have to manually increase this number in a rundeck job I think a timestamp might be a good intermittent solution before our switch to Conan and Pipelines.

Special care should be taken to make sure that timestamp is used for all three OSes. that requires us to set it in Rundeck and not here.

@evtrados Can you make sure that all three build jobs on Rundeck use the same timestamp ad the CURA_VERSION_BUILD